corehq/apps/tzmigration/api.py

```diff
@@ -0,0 +1,81 @@
+from django.conf import settings
+from corehq.apps.tzmigration.exceptions import TimezoneMigrationProgressError
+from corehq.util.quickcache import skippable_quickcache
+from corehq.util.soft_assert import soft_assert
+from corehq.util.view_utils import get_request
+from models import TimezoneMigrationProgress, MigrationStatus
+
+
+def set_migration_started(domain):
+    progress, _ = TimezoneMigrationProgress.objects.get_or_create(pk=domain)
+    if progress.migration_status == MigrationStatus.NOT_STARTED:
+        progress.migration_status = MigrationStatus.IN_PROGRESS
+        progress.save()
+        # reset cache
+        get_migration_status(domain, strict=True)
+    else:
+        raise TimezoneMigrationProgressError(
+            'Cannot start a migration that is already in state {}'
+            .format(progress.migration_status)
+        )
+
+
+def set_migration_complete(domain):
+    progress, _ = TimezoneMigrationProgress.objects.get_or_create(pk=domain)
+    if progress.migration_status != MigrationStatus.COMPLETE:
+        progress.migration_status = MigrationStatus.COMPLETE
+        progress.save()
+        # reset cache
+        get_migration_status(domain, strict=True)
+
+
+def get_migration_complete(domain):
+    return get_migration_status(domain) == MigrationStatus.COMPLETE
+
+
+@skippable_quickcache(['domain'], skip_arg='strict')
+def get_migration_status(domain, strict=False):
+    progress, _ = TimezoneMigrationProgress.objects.get_or_create(pk=domain)
+    return progress.migration_status
+
+
+def phone_timezones_have_been_processed():
+    """
+    The timezone data migration happening some time in Apr-May 2015
+    will shift all phone times (form.timeEnd, case.modified_on, etc.) to UTC
+    so functions that deal with converting to or from phone times
+    use this function to decide what type of timezone conversion is necessary
+
+    """
+    if settings.UNIT_TESTING:
+        override = getattr(
+            settings, 'PHONE_TIMEZONES_HAVE_BEEN_PROCESSED', None)
+        if override is not None:
+            return override
+    return (_get_migration_status_from_threadlocals()
+            == MigrationStatus.COMPLETE)
+
+
+def phone_timezones_should_be_processed():
+    if settings.UNIT_TESTING:
+        override = getattr(
+            settings, 'PHONE_TIMEZONES_SHOULD_BE_PROCESSED', None)
+        if override is not None:
+            return override
+    return _get_migration_status_from_threadlocals() in (
+        MigrationStatus.IN_PROGRESS, MigrationStatus.COMPLETE)
+
+
+def _get_migration_status_from_threadlocals():
```

Comment:
Obvious downside is it relies on a function being processed in the same thread as the triggering request was made, which makes it harder to refactor something to be done asynchronously (with e.g. celery) or by a separate little service if/when we start going that route.